features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/SessionReplayFeature.kt

```diff
@@ -193,30 +227,52 @@ internal class SessionReplayFeature(
                 InternalLogger.Target.USER,
                 { EVENT_MISSING_MANDATORY_FIELDS }
             )
-            return
+            return null
         }
 
-        if (currentRumSessionId.get() == sessionId) {
-            // we already handled this session
-            return
+        return SessionData(keepSession, sessionId)
+    }
+
+    @Suppress("ReturnCount")
+    private fun shouldHandleEvent(sessionData: SessionData): Boolean {
+        if (currentRumSessionId.get() != sessionData.sessionId) {
+            // sessionId changed so resample
+            isSessionSampledIn.set(rateBasedSampler.sample())
+        } else {
+            if (!recordingStateChanged.get()) {
+                // sessionId was already seen and recordingState did not change
+                return false
+            }
         }
 
         if (!checkIfInitialized()) {
-            return
+            return false
+        }
+
+        val isSampledIn = sessionData.keepSession && isSessionSampledIn.get()
+
+        if (!isSampledIn) {
+            if (shouldRecord.getAndSet(false)) {
+                sdkCore.internalLogger.log(
+                    InternalLogger.Level.INFO,
+                    InternalLogger.Target.USER,
+                    { SESSION_SAMPLED_OUT_MESSAGE }
+                )
+            }
         }
 
-        if (keepSession && rateBasedSampler.sample()) {
+        recordingStateChanged.set(false)
```

Comment:
why is this flag handled here ? shouldn't it be handled in the `handleRecording` ?